### PR TITLE
Shifts to Orders API (Auto Capture)

### PR DESCRIPTION
**Pull request by mayankamencherla**
Check and merge @captn3m0 @73sl4

### CODE REVIEW

File: razorpay-payments.php [hunk 1]
```diff
@@ -141,12 +143,38 @@ public function generate_razorpay_form($order_id)
...
+
             $order = new WC_Order($order_id);
 
             $redirect_url = get_site_url() . '/?wc-api=' . get_class($this);
 
             $productinfo = "Order $order_id";
 
+            $api = new Api($this->key_id, $this->key_secret);
+
+            if($this->payment_action === 'authorize'){
```
Comment (captn3m0) [inline razorpay-payments.php]: switch case would be better

File: razorpay-payments.php [hunk 2]
```diff
@@ -141,12 +143,38 @@ public function generate_razorpay_form($order_id)
...
+                $data = array(
+                'receipt'         => $order_id,
```
Comment (captn3m0) [inline razorpay-payments.php]: alignment (give proper spacing here)

File: razorpay-payments.php [hunk 3]
```diff
@@ -141,12 +143,38 @@ public function generate_razorpay_form($order_id)
...
+                'amount'          => (int) ($order->order_total * 100),
+                'currency'        => get_woocommerce_currency(),
+                'payment_capture' => 1
+                );
+            }
+            
+            $razorpay_order = $api->order->create($data);
+            
+            $woocommerce->session->set('razorpay_order_id',$razorpay_order['id']);
```
Comment (captn3m0) [inline razorpay-payments.php]: space after comma

File: razorpay-payments.php [hunk 4]
```diff
@@ -292,44 +324,40 @@ function check_razorpay_response()
...
-                        
-                        $payment = $payment->capture(array('amount'=>$amount)); 
+                        $razorpay_order_id = $woocommerce->session->get('razorpay_order_id');
+                        $razorpay_signature = $_REQUEST['razorpay_signature'];
                         
-                        $expectedStatus = 'captured';
+                        $signature = hash_hmac('sha256', $razorpay_order_id . '|' . $razorpay_payment_id, $key_secret);
+
+                        if( hash_equals($signature , $razorpay_signature) ){
+                            $status = 'captured';
```
Comment (captn3m0) [inline razorpay-payments.php]: $status is never used as a string anywhere. Create a boolean `$captured` instead.

File: razorpay-payments.php [hunk 5]
```diff
@@ -292,44 +324,40 @@ function check_razorpay_response()
...
+                        $status = 'authorized';
                     }
                     else
                     {
-                        $payment = $api->payment->fetch($razorpay_payment_id);
-                        $amount = $payment->amount;
-                        
-                        $payment = $payment->capture(array('amount'=>$amount)); 
+                        $razorpay_order_id = $woocommerce->session->get('razorpay_order_id');
+                        $razorpay_signature = $_REQUEST['razorpay_signature'];
```
Comment (captn3m0) [inline razorpay-payments.php]: - use $_POST over $_REQUEST here. Very slight advantage in terms of knowing where the input is coming from
- https://stackoverflow.com/questions/1149118/does-request-have-security-problem

Maybe implement a validation for the input values here.

File: razorpay-payments.php [hunk 6]
```diff
@@ -292,44 +324,40 @@ function check_razorpay_response()
...
                         
-                        $expectedStatus = 'captured';
+                        $signature = hash_hmac('sha256', $razorpay_order_id . '|' . $razorpay_payment_id, $key_secret);
+
+                        if( hash_equals($signature , $razorpay_signature) ){
```
Comment (captn3m0) [inline razorpay-payments.php]: extra spacing inside if, put braces in newline

File: razorpay-payments.php (hunk 4 above)
Comment (mayankamencherla) [inline razorpay-payments.php]: @captn3m0 done.

File: razorpay-payments.php [hunk 7]
```diff
@@ -141,12 +143,41 @@ public function generate_razorpay_form($order_id)
         {
             global $woocommerce;
 
+            // currently in the form only razorpay payment ID is there. But with the addition of order ID and signature, this should be straight forward
```
Comment (captn3m0) [inline razorpay-payments.php]: try to limit comments to 80 characters. Split into two lines.

File: razorpay-payments.php [hunk 8]
```diff
@@ -141,12 +143,41 @@ public function generate_razorpay_form($order_id)
...
 
             $productinfo = "Order $order_id";
 
+            $api = new Api($this->key_id, $this->key_secret);
+
+            switch($this->payment_action)
+            {
+                case 'authorize':
+                    $data = array(
+                      'receipt' => $order_id,
```
Comment (captn3m0) [inline razorpay-payments.php]: Make a wrapper function called `getOrderCreationData` or something similar.

File: razorpay-payments.php [hunk 9]
```diff
@@ -281,9 +316,9 @@ function check_razorpay_response()
 
             $order_id = $woocommerce->session->get(self::SESSION_KEY);
 
-            if ($order_id  and !empty($_REQUEST['razorpay_payment_id']))
```
Comment (captn3m0) [inline razorpay-payments.php]: This function is getting too long. Can we split it into smaller pieces?

Comment (mayankamencherla) [inline razorpay-payments.php]: Working on it @captn3m0

**Comment by mayankamencherla** [comment]
Done @captn3m0

File: razorpay-payments.php [hunk 10]
```diff
@@ -175,7 +175,7 @@ public function generate_razorpay_form($order_id)
 
             $json = json_encode($razorpay_args);
 
-            $html = $this->generate_order_form($redirect_url,$json);
+            $html = $this->generate_order_form($redirect_url,$json,$order_id);
```
Comment (captn3m0) [inline razorpay-payments.php]: space after commas

File: razorpay-payments.php (hunk 8 above)
Comment (captn3m0) [inline razorpay-payments.php]: @mayankamencherla This is pending.

File: razorpay-payments.php [hunk 11]
```diff
@@ -292,44 +348,39 @@ function check_razorpay_response()
...
-                        $amount = $payment->amount;
+                        $razorpay_order_id = $woocommerce->session->get('razorpay_order_id');
+                        $razorpay_signature = $_POST['razorpay_signature'];
                         
-                        $payment = $payment->capture(array('amount'=>$amount)); 
-                        
-                        $expectedStatus = 'captured';
+                        $signature = hash_hmac('sha256', $razorpay_order_id . '|' . $razorpay_payment_id, $key_secret);
+
+                        if (hash_equals($signature , $razorpay_signature))
```
Comment (captn3m0) [inline razorpay-payments.php]: Hash_equals has a dependency on PHP 5.6. Find a polyfill that you can use.

File: razorpay-payments.php [hunk 12]
```diff
@@ -210,8 +265,9 @@ public function generate_razorpay_form($order_id)
       var successMsg = document.getElementById('msg-razorpay-success');
       successMsg.style.display = "block";
 
-      document.getElementById('razorpay_payment_id').value =
-        payment.razorpay_payment_id;
+      document.getElementById('razorpay_payment_id').value = payment.razorpay_payment_id;
+      document.getElementById('razorpay_order_id').value = payment.razorpay_order_id;
```
Comment (captn3m0) [inline razorpay-payments.php]: Why submit `razorpay_order_id` if you are picking it up from session?

File: razorpay-payments.php [hunk 13]
```diff
@@ -292,44 +348,39 @@ function check_razorpay_response()
...
                 $api = new Api($key_id, $key_secret);
-
+                
                 try
                 {
                     if ($this->payment_action === 'authorize')
                     {   
                         $payment = $api->payment->fetch($razorpay_payment_id);
-            
-                        $expectedStatus = 'authorized';
```
Comment (captn3m0) [inline razorpay-payments.php]: Put an assert here on the payment amount matching the woocommerce order amount.

File: razorpay-payments.php [hunk 14]
```diff
@@ -292,44 +348,39 @@ function check_razorpay_response()
...
-                    if ($payment->status === $expectedStatus)
+                    if ($captured)
                     {
                         $success = true;
                     }
-                    else
-                    {
+
+                    else{
                         $success = false;
```
Comment (captn3m0) [inline razorpay-payments.php]: The else block will execute on action=authorize and raise an error.

File: razorpay-payments.php [hunk 15]
```diff
@@ -173,6 +226,8 @@ public function generate_razorpay_form($order_id)
 <form name='razorpayform' action="$redirect_url" method="POST">
     <input type="hidden" name="merchant_order_id" value="$order_id">
```
Comment (captn3m0) [inline razorpay-payments.php]: I don't see `merchant_order_id` field being used anywhere after form submission either. Drop it?

File: razorpay-payments.php [hunk 16]
```diff
@@ -160,11 +169,55 @@ public function generate_razorpay_form($order_id)
...
               'notes' => array(
                 'woocommerce_order_id' => $order_id
-              )
+              ),
+              'order_id' => $razorpay_order['id']
             );
 
             $json = json_encode($razorpay_args);
 
+            $html = $this->generate_order_form($redirect_url,$json,$order_id);
```
Comment (captn3m0) [inline razorpay-payments.php]: space after comma.

File: razorpay-payments.php (hunk 14 above)
Comment (mayankamencherla) [inline razorpay-payments.php]: I'm checking to see if payment['amount'] === $amount and setting success = true in that case. That should fix the error. 2 birds with 1 stone. Sounds good? @captn3m0

File: razorpay-payments.php (hunk 8 above)
Comment (mayankamencherla) [inline razorpay-payments.php]: @captn3m0 Done. I didn't push my latest commit. It shows on webhooks.

File: razorpay-payments.php (hunk 14 above)
Comment (captn3m0) [inline razorpay-payments.php]: `$payment->amount === $amount` sounds good. I think $amount descends from the WC_Order, which makes sense.

We still need the status check though. The payment amounts might match, but we need the payment to be captured.

File: razorpay-payments.php (hunk 8 above)
Comment (captn3m0) [inline razorpay-payments.php]: Can you push the commit here? Cherry pick it on this branch and push?

File: razorpay-payments.php [hunk 17]
```diff
@@ -294,41 +347,32 @@ function check_razorpay_response()
...
-                            $error = $payment['error_code'] . ": " . $payment['error_description'];
+                            $success = true;
                         }
+
                         else
                         {
-                            $error = 'RAZORPAY_ERROR: Invalid Response';
+                            $success = false;
+
+                            $error = "PAYMENT_ERROR = Payment failed";
```
Comment (captn3m0) [inline razorpay-payments.php]: Use a colon, instead of `=`

File: razorpay-payments.php [hunk 18]
```diff
@@ -147,6 +147,15 @@ public function generate_razorpay_form($order_id)
 
             $productinfo = "Order $order_id";
 
+            $api = new Api($this->key_id, $this->key_secret);
```
Comment (captn3m0) [inline razorpay-payments.php]: We should ideally have a try/catch around all network calls.

**Comment by captn3m0** [comment]
@mayankamencherla can you squash all your commits to a single commit?

**Pull request closed without merging**